packages/foundry/src/vault/adapter/aave/aaveV2/AaveV2Adapter.sol

```diff
@@ -0,0 +1,162 @@
+// SPDX-License-Identifier: GPL-3.0
+// Docgen-SOLC: 0.8.15
+
+pragma solidity ^0.8.15;
+
+import { AdapterBase, IERC20, IERC20Metadata, SafeERC20, ERC20, Math, IStrategy, IAdapter } from "../../abstracts/AdapterBase.sol";
+import { WithRewards, IWithRewards } from "../../abstracts/WithRewards.sol";
+import { ILendingPool, IAaveMining, IAToken } from "./IAaveV2.sol";
+
+/**
+ * @title   AaveV2 Adapter
+ * @author  amatureApe
+ * @notice  ERC4626 wrapper for AaveV2 Vaults.
+ *
+ * An ERC4626 compliant Wrapper for https://github.com/aave/protocol-v2/blob/master/contracts/protocol/lendingpool/LendingPool.sol.
+ * Allows wrapping AaveV2 aTokens with or without an active Liquidity Mining.
+ * Allows for additional strategies to use rewardsToken in case of an active Liquidity Mining.
+ */
+
+contract AaveV2Adapter is AdapterBase, WithRewards {
+  using SafeERC20 for IERC20;
+  using Math for uint256;
+
+  string internal _name;
+  string internal _symbol;
+
+  // @notice The Aave aToken contract
+  IAToken public aToken;
+
+  // @notice The Aave liquidity mining contract
+  IAaveMining public aaveMining;
+
+  // @notice Check to see if Aave liquidity mining is active
+  bool public isActiveMining;
+
+  // @notice The address that will receive the liquidity mining rewards if LM is active
+  address public rewardsRecipient;
+
+  // @notice The Aave LendingPool contract
+  ILendingPool public lendingPool;
+
+  /*//////////////////////////////////////////////////////////////
+                            CONSTRUCTOR
+    //////////////////////////////////////////////////////////////*/
+
+  /**
+   * @notice Initialize a new AaveV2 Adapter.
+   * @param adapterInitData Encoded data for the base adapter initialization.
+   * @param aaveInitData Encoded data for the beefy adapter initialization.
+   * @dev `_aToken` - The underlying asset supplied to and wrapped by Aave.
+   * @dev `_lendingPool` - The lending pool.
+   * @dev `_aaveMining` - An optional liquidity mining contract to boost yield.
+   * @dev This function is called by the factory contract when deploying a new vault.
+   */
+  function initialize(
+    bytes memory adapterInitData,
+    address,
+    bytes memory aaveInitData
+  ) public {
+    (address _aToken, address _lendingPool, address _aaveMining, address _rewardsRecipient) = abi.decode(
+      aaveInitData,
+      (address, address, address, address)
+    );
+    __AdapterBase_init(adapterInitData);
+
+    _name = string.concat("Popcorn AaveV2", IERC20Metadata(asset()).name(), " Adapter");
+    _symbol = string.concat("popB-", IERC20Metadata(asset()).symbol());
+
+    aToken = IAToken(_aToken);
+    lendingPool = ILendingPool(_lendingPool);
+    aaveMining = IAaveMining(_aaveMining);
+    // Method implemented manually inputs mining contract and then checks for emission.
+    // Commented method below queries the aToken for contract.
+    // **IAaveIncentivesController aaveMining = aToken.getIncentivesController();**
```

Comment:
This should probably have a check if you want to go for the programmatic route to make sure there is a mining contract. Also, the type IAaveIncentivesController is declared as IAaveMining in the interfaces so that needs to be changed as well.

Reply:
Agreed!